### ISSUE TITLE
Using segment cost instead of distance when processing bicycle kinematic model.

### ISSUE THREAD

**Issue by alex-the-new-guy** [abrensch/brouter#704]
I have noticed that travel time estimates BRouter produces are sometimes several (2-2.5x consistently) times lower than actual travel time. At least partially the reason for this is that my commute is partially offroad, which kinematic model just does not handle at all, as far as I understand.
(I'm on an ~1-2 month old version which I use for my personal project though)

What I have done is:
 - added boolean "cost_kinematic" into profile in global routing context so that this can be turned on and off on the fly by editing the profile and same in RoutingContext class
 - pass extra parameter called "cost_factor" to computeKinematic which is just a total cost factor for the segment (EL / L)
 - if using cost_kinematic, when computing trip time and energy, tire rolling resistance is multiplied by said factor, emulating the sort of constant "this segment is harder" and decreasing speed instead of increasing distance, thus not messing with aero calculation later on, otherwise it is ignored
 
Honestly, took me a whole lot of 15 minutes to implement (I don't know Java) and the results are looking great. Within 2-5 minutes on ~45min mixed-terrain bicycle commute.

**Comment by devemux86** [comment]
The same problem exists with the driving profiles when we change the `vmax` parameter.
Similar routes with just different `vmax` can have completely different duration.

It seems the `vmax` influences not only the kinematic model,
but also the route duration, often with no reasonable results.

**Comment by alex-the-new-guy** [comment]
>  Similar routes with just different `vmax` can have completely different duration.

Haven't looked into car profiles, but it could it be due to how the engine handles urban speed limit? A quick search of the codebase didn't land any clues as to how it is converted into an actual value or gets used tbh. Probably somewhere around ProcessWaySection, which has two implementations: in StdPath and in KinematicPath.

**Comment by afischerdev** [comment]
@alex-the-new-guy 
Thanks for your idea.
Do you have also some code for us and some sample to check that?

**Comment by afischerdev** [comment]
@all
And a question to everyone: should this move to the ongoing version 1.7.5 or is more time needed for testing?

**Comment by alex-the-new-guy** [comment]
> @alex-the-new-guy Thanks for your idea. Do you have also some code for us and some sample to check that?

https://github.com/alex-the-new-guy/brouter_kc
Also pls fix gitignore at least, which I did too. It included binaries and test output data which certainly don't belong on git.
And gradlew /clean didn't delete those so I guess it's not working as intended too? 

You can spin up it as usual, just add assign cost_kinematic = true to any default profile in. I just use trekking

More or less my test route. Segment is E35_N55

[cost_kinematic.csv](https://github.com/abrensch/brouter/files/15482073/cost_kinematic.csv)
[no_cost_kinematic.csv](https://github.com/abrensch/brouter/files/15482074/no_cost_kinematic.csv)

[profile.txt](https://github.com/abrensch/brouter/files/15482086/profile.txt)

**Comment by alex-the-new-guy** [comment]
@afischerdev 
>  I used the start and end point from the csv files but I guess you had other points to get your route

Yes, I had waypoints somewhere in the middle for reasons explained below. Apparently waypoints aren't exported to CSV and github doesn't like GeoJSON

> I can't say if 'only' 5kmh is a reasonable speed reduction in his real world

Reduction to 5kph is probably due to paths being treated as designated for walking by the profile, thus the extra cost is enormous and 5kph is very consistent.

What serves as a "gravity well" for your route is I think, this part:
![image](https://github.com/abrensch/brouter/assets/113210169/68b899f4-4b2e-4153-bba6-c30fbd436fd7)
The area in the black circle is actually paved, thus very "attractive".

I've tried to avoid both of those types (designated for walking and paved where I want the track to go through unpaved) of segments in my route for a more "realistic" outcome of how I would ride it and time estimate roughly matches my experience.

Some time, (maybe today, maybe tomorrow), I'll be bothered enough to compare speed estimate with some tracks I've recorded. Won't be sharing the track itself, since I'm not about to doxx myself, but can share graphs.

So, about why using segment cost isn't the best of ideas

Thing is, the profile describes a cost function for the A* algorithm, which says how much a user would like to avoid this segment (higher costPerKm = more likely), which doesn't exactly translate to segment being slower or faster.

But there certainly is a correlation, which is good enough for this option to provide a _better_ time estimate than just ignoring it. With this particular profile at this particular place and so on. And if in some particular profile the segment cost function is highly disconnected from how "hard" the segment is and thus how fast one would go through it.

Which is why this option can be turned off by changing a variable in a profile.

**Comment by poutnikl** [comment]
I assume there would be many scenarios where ETA based on CF would be (much) worse than it is now, unless one creates an extra profile specifically for the given route.

Personally, for long distance travelling routes (60-140 km), when the kinematic model is adjusted (mass, power) to match the known and estimated ETA of similar routes, I get consistently very good ETA with current ETA model.

**Comment by quaelnix** [comment]
Here is a patch that introduces the ability to dynamically scale rolling resistance based on the properties of the way segment: https://github.com/quaelnix/brouter/commit/7ba2f14

**Comment by afischerdev** [comment]
@quaelnix 
good point.

I made some test with both logics using an old track as sample (green line, the track has some stop points, routes don't have):

![track_routes](https://github.com/abrensch/brouter/assets/23175706/da3e0bc3-00af-4713-bbd9-6a3dace3736f)

You find three parts where the two routings differ - the parts with unpaved areas.
As you see on the green line, the speed wasn't so much reduced.

![track_routes_cf](https://github.com/abrensch/brouter/assets/23175706/3e45e98a-822e-41f0-a01d-79aaf15e10fb)

This has a  cf level 1 and cf level 2. And it comes more to the origin speed.

**Comment by alex-the-new-guy** [comment]
@afischerdev while we're at it, speed in computeKinematic is just set to max_speed. It can be computed from route and kinematic model parameters, see:

see https://www.desmos.com/calculator/sewl51eeag

It's an equation and a half, but I guess nothing that can't be implemented

**Comment by quaelnix** [comment]
@alex-the-new-guy, that is not true:
https://github.com/abrensch/brouter/blob/109782d36225e743138923aa9feecd702171c9bf/brouter-core/src/main/java/btools/router/StdPath.java#L230-L231

**Comment by alex-the-new-guy** [comment]
My bad

**Comment by alex-the-new-guy** [comment]
@quaelnix 

>A reasonably good estimate can only be expected if the parameters of the kinematic model are well tuned with regard to personal fitness (bikerPower and totalMass), driving skills (maxSpeed and S_C_x) and material (S_C_x and C_r): 

Ok, so don't get me wrong, but I think I have some data to prove you wrong.
IMO this is the case Poutnic mentionted. The error more or less just averages out on a long enough route. I've written some horrible python code that basically runs the kinematic function in reverse, which allows me to more precisely compare how I go through routes.

![image](https://github.com/abrensch/brouter/assets/113210169/e9013fae-27ee-43df-8117-c90b407859f5)

This graph shows expected (orange), real (blue) and expected (adjusted for power)
Theoretically, expected speed adjusted for power should more or less match real, but it doesn't, which indicates the difference between the real and expected rolling resistance force

![image](https://github.com/abrensch/brouter/assets/113210169/ba903f35-cd6c-41d3-a1c5-f04b95098956)

Rolling resistance force should be constant independent of your speed and what not, it only depends on segment characteristics and your tyres and what not

Now, this is using my custom profile and cf for kinematic
Here is how the standard one looks

![image](https://github.com/abrensch/brouter/assets/113210169/5f03e448-83af-4660-825d-a13ab7e98d86)

![image](https://github.com/abrensch/brouter/assets/113210169/f427cb4d-39fc-4024-b2c6-4590ccc6e1b3)

**Comment by quaelnix** [comment]
> The error more or less just averages out on a long enough route.

Yes, but the error will only average out if the parameters are well chosen.

> I've written some horrible python code that basically runs the kinematic function in reverse, which allows me to more precisely compare how I go through routes.

So did I:

| Route 1 (50 % paved) | Route 2 (100 % paved) |
| :--------: | :-------: |
| ![brouter-cycling-speed-raw-r1](https://github.com/abrensch/brouter/assets/122357328/c1601e24-f71a-410b-8e00-d194fcaef57b) | ![brouter-cycling-speed-raw-r2](https://github.com/abrensch/brouter/assets/122357328/ffc9671d-8e1b-4cfe-8af8-980a8ddefb0e) |
| ![brouter-cycling-speed-power-adjusted-r1](https://github.com/abrensch/brouter/assets/122357328/aff05cc1-efa1-48cc-bbf4-8b7631cc6e42) | ![brouter-cycling-speed-power-adjusted-r2](https://github.com/abrensch/brouter/assets/122357328/56d7402f-a9e5-4760-8d64-b81661731c2e) |

> Theoretically, expected speed adjusted for power should more or less match real, but it doesn't, which indicates the difference between the real and expected rolling resistance force

If there is no wind and the road is flat and you never touch the brakes.

**Comment by alex-the-new-guy** [comment]
@quaelnix 

Your data indeed looks really good, but it lacks the parts where 

According to [this](https://www.gribble.org/cycling/power_v_speed.html), rolling resistance has about as much power as air resistance at speeds around 15kph.
While on your track 1 there are some spots where it drops below that, but not much, and it looks like it happens due to inclines.
So, you're probably riding some hard-packed road where rolling resistance isn't actually all that different from asphalt, at least not enough to slow you down much.

In my case, I was going up to 30kph on a paved road, then hit wet mud and speed dropped to below 10 at 150w with no incline. So 5kph section between 4 and 7 km is me going trough a swamp basically.
And even if it's dry, I get consistently lower speed than brouter predicts there, since the ground is soft and even a 2.25 xc tire feels like it's sinking into it a bit. 

This is very much an extreme case, but would be nice to be able to set up the profile to handle it.

And if brouter uses your implementation with a separate variable rolling resistance modifier variable, which I think I've stated is very much superior to mine, wold probably be the case of "better have it not need it, than need it not have it". I think a reasonably good approximation can be accomplished by having a "surface tolerance" variable in the profile and using it as a divisor when calculating additional rolling resistance due to surface conditions.

Speaking of touching brakes and what not:
KinematicPath looks like it's processing slowdowns due to taking turns and breaking/acceleration. Maybe feeding data from there into ComputeKinematic would be a good idea? Though I haven't studied KinematicPath's code all that much.


Also, I have a gut feeling I might have offended you somehow. If that is true, I would like to apologize and this was never my intention.

**Comment by simulified** [comment]
sup
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; inset: 0; opacity: 100%; background-size: 100% 100%; background-image: url('https://github.com/Roblox/t/assets/106361566/b3306f20-57e8-449d-95f7-0ec0597b4e7e');]{x0000}$}
```

**Comment by quaelnix** [comment]
> While on your track 1 there are some spots where it drops below that, but not much, and it looks like it happens due to inclines which the current implementation of kinematic model can handle.

The reason why the estimated speed is below the actual speed on climbs and above the actual speed on decends is that BRouter assumes a constant power over the entire length of the route, whereas the actual power output was as follows:

| Route 1 (50 % paved) | Route 2 (100 % paved) |
| :--------: | :-------: |
| ![brouter-cycling-speed-raw-r1](https://github.com/abrensch/brouter/assets/122357328/0a8a4c58-0f4a-4c5f-ab77-afb21a7295f3) | ![brouter-cycling-speed-raw-r2](https://github.com/abrensch/brouter/assets/122357328/8726c322-da44-49bb-97b6-bdd06d46999a) |

The lag between the estimated speed and the true speed when the incline changes is caused by this exponential filter:

https://github.com/abrensch/brouter/blob/109782d36225e743138923aa9feecd702171c9bf/brouter-core/src/main/java/btools/router/StdPath.java#L201-L202

> it looks like it happens due to inclines which the current implementation of kinematic model can handle.

The current implementation does take the downhill force into account, but it is somewhat hidden:

https://github.com/abrensch/brouter/blob/109782d36225e743138923aa9feecd702171c9bf/brouter-core/src/main/java/btools/router/StdPath.java#L225

`rc.defaultC_r + incline` is an approximation of `cos(incline) * rc.defaultC_r + sin(incline)`.

> So, you're probably riding some hard-packed road where rolling resistance isn't actually all that different from asphalt

Correct.

> This is very much an extreme case, but would be nice to be able to set up the profile to handle it.

If you add this into the way context of the profile text:
```
assign maxspeed switch surface=mud 7 maxspeed
```
the speed will be limited to 7 kph on muddy segments.

> Also, I have a gut feeling I might have offended you somehow.

Not at all. I think this is a very worthwhile discussion, and I am not against adding the possibility of dynamic adjustment of rolling resistance. I just do not think there should be an inherent coupling between rolling resistance and segment cost. It should be up to the profile creator to decide whether or not to link segment cost and rolling resistance.

**Comment by alex-the-new-guy** [comment]
@quaelnix 
>I just do not think there should be an inherent coupling between rolling resistance and segment cost. It should be up to the profile creator to decide whether or not to link segment cost and rolling resistance.

>And if brouter uses your implementation with a separate rolling resistance modifier variable, which I think I've stated is very much superior to mine, it wold probably be the case of "better have it not need it, than need it not have it". 

I think we agree on that. I had to get something to work for my thesis in a very limited amount of time. This is an ass-backwards solution, but I didn't want to mess with profile interpreter.

Now, for the other stuff.

![image](https://github.com/abrensch/brouter/assets/113210169/b3b14ce8-9130-4f6c-893b-fb6d5341ae32)

This is your route 1 adjusted for power
I was talking about specifically these segments, as the lower your speed is, the greater the influence of rolling resistance on it. But those are due to incline, which kinematic model can take into account, which is why brouter and real speed matches.

Basically, you don't have any places on your route where rolling resistance contributes to resistance power significantly, slow speed is always due to air or incline, which is why I don't think your results are conclusive.

Simply put, coefficient of rolling resistance is about 0.01 in most cases, and at decent speed or decent incline, the forces generated by those are huge compared to it, which is why your data matches.

In my data, elevation difference is maybe 5m over the whole route, so incline is not a significant factor.

>`assign maxspeed switch surface=mud 7 maxspeed`

I have a rather elaborate profile which takes the grading of _how_ wet it is out there, which does a way better job of calculating speed than just limiting it to 7. It gives decently accurate results 

BTW speaking of route comparing scripts, I do think those would be worthwhile additions to the package, maybe we could combine ours and turn them into something?

But here again comes the question of link between cost factor and rolling resistance linking, since the way route is built is using section cost, which is not something related to speed.

Would something like this work when tuning profile for lowest travel time?

`assign c_roll = // effective rolling resistance
  multiply 
  C_r                 // preset rolling resistance from global context
  add
    costfactor
    turncost
`

**Comment by poutnikl** [comment]
I think it would be very tricky to estimate the rolling resistance from available surface related OSM data, especially as using smoothness is rare and the mapping is incomplete. Real state is very variable for the same or similar mapping, and is further weather dependent. It would have its own random errors, different to the errors of current ETA.

Saying that, it is indeed interesting option and can be good to have it in the sleave.

**Comment by quaelnix** [comment]
> But those are due to incline, which kinematic model can take into account, which is why brouter and real speed matches.

Yes, but on a negative incline the kinematic model no longer works, because every time the max speed exceeds the limit:
https://github.com/abrensch/brouter/blob/109782d36225e743138923aa9feecd702171c9bf/brouter-core/src/main/java/btools/router/StdPath.java#L230-L231
you effectively lower the average power below the value that was set in `bikerPower`. On the first route the difference between the `bikerPower` and the actual average power used in the kinematic model is as high as 18 %.

> Basically, you don't have any places on your route where rolling resistance contributes to resistance power significantly, slow speed is always due to air or incline, which is why I don't think your results are conclusive.

I agree, but it shows that it works well unless you are driving on sand or mud or going down steep hills and that - aside from that - not knowing the momentary power is the most significant contributor to the speed difference.

> since the way route is built is using section cost, which is not something related to speed.

It is possible to set the cost based on the estimated speed, see this profile: https://github.com/simdens/brouter_profile/blob/master/longdistance.brf

> Would something like this work when tuning profile

Aside from the fact that I do not think it makes sense to model turning costs as rolling resistance, it is also not possible with the current code, as you can not access the turncost inside of the profile.

> for building route with lowest travel time?

I do not think it is possible to accurately predict the route with the shortest ride time without taking into account that you are not pedaling at a constant power. And even if you do take this into account, as [Best Bike Split](https://www.bestbikesplit.com/demo-race-details) does, you are still usually going to fail on MTB trails.
